pkg/cmd/roachtest/tests/failure_injection.go

```diff
@@ -0,0 +1,240 @@
+	// directly.
+	failureMode, err := fr.GetFailureMode(c.MakeNodes(), t.failureName, l, c.IsSecure())
+	if err != nil {
+		return err
+	}
+	if err = failureMode.Setup(ctx, l, t.args); err != nil {
+		return err
+	}
+	if err = failureMode.Inject(ctx, l, t.args); err != nil {
+		return err
+	}
+
+	// Allow the failure to take effect.
+	if err = failureMode.WaitForFailureToPropagate(ctx, l, t.args); err != nil {
+		return err
+	}
+
+	if err = t.validateFailure(ctx, l, c); err != nil {
+		return err
+	}
+	if err = failureMode.Restore(ctx, l, t.args); err != nil {
+		return err
+	}
+
+	// Allow the cluster to return to normal.
+	if err = failureMode.WaitForFailureToRestore(ctx, l, t.args); err != nil {
+		return err
+	}
+
+	if err = t.validateRestore(ctx, l, c); err != nil {
+		return err
+	}
+	if err = failureMode.Cleanup(ctx, l, t.args); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *failureSmokeTest) noopRun(
+	ctx context.Context, l *logger.Logger, c cluster.Cluster, _ *failures.FailureRegistry,
+) error {
+	if err := t.validateFailure(ctx, l, c); err == nil {
+		return errors.New("no failure was injected but validation still passed")
+	}
+	if err := t.validateRestore(ctx, l, c); err != nil {
+		return errors.Wrapf(err, "no failure was injected but post restore validation still failed")
+	}
+	return nil
+}
+
+var bidirectionalNetworkPartitionTest = failureSmokeTest{
+	testName:    "bidirectional network partition",
+	failureName: failures.IPTablesNetworkPartitionName,
+	args: failures.NetworkPartitionArgs{
+		Partitions: []failures.NetworkPartition{{
+			Source:      install.Nodes{1},
+			Destination: install.Nodes{3},
+			Type:        failures.Bidirectional,
+		}},
+	},
+	validateFailure: func(ctx context.Context, l *logger.Logger, c cluster.Cluster) error {
+		blocked, err := roachtestutil.CheckPortBlocked(ctx, l, c, c.Nodes(1), c.Nodes(3), "{pgport:3}")
+		if err != nil {
+			return err
+		}
+		if !blocked {
+			return fmt.Errorf("expected connections from node 1 to node 3 to be blocked")
+		}
+
+		blocked, err = roachtestutil.CheckPortBlocked(ctx, l, c, c.Nodes(1), c.Nodes(2), "{pgport:2}")
+		if err != nil {
+			return err
+		}
+		if blocked {
+			return fmt.Errorf("expected connections from node 1 to node 2 to not be blocked")
+		}
+		return nil
+	},
+	validateRestore: func(ctx context.Context, l *logger.Logger, c cluster.Cluster) error {
+		blocked, err := roachtestutil.CheckPortBlocked(ctx, l, c, c.Nodes(1), c.Nodes(3), "{pgport:3}")
+		if err != nil {
+			return err
+		}
+		if blocked {
+			return fmt.Errorf("expected connections from node 1 to node 3 to not be blocked")
+		}
+		return err
+	},
+}
+
+var asymmetricNetworkPartitionTest = failureSmokeTest{
+	testName:    "asymmetric network partition",
+	failureName: failures.IPTablesNetworkPartitionName,
+	args: failures.NetworkPartitionArgs{
+		Partitions: []failures.NetworkPartition{{
+			Source:      install.Nodes{1},
```

Comment:
Granted it's a "smoke test", I wonder if it would strengthen it by randomizing a pair of nodes instead of hardcoding.

Reply:
Done.